### PR TITLE
#17047 Airbyte CDK: Improve error for returning non-iterable from connectors parse_response

### LINKED ISSUE #17047: Airbyte CDK: Improve error for returning non-iterable from connectors parse_response
## Environment
- **Airbyte version**:  0.40.7
- **OS Version / Instance**: macOS
- **Deployment**: Docker
- **Source Connector and version**: Custom python HTTP source
- **Step where error happened**: Running `python3 main.py read` while developing the connector

## Current Behavior
If the `parse_response` method of the stream does not return an iterator, this leads to a fairly cryptic pydantic error message, instead of specyfing that an iterator is expected. As `response.json()` is type hinted with `Any`, my IDE did not catch this type violation for me and I was quite confused.

## Expected Behavior
The error should tell me that the iterator returned from my `parse_response` function is incompatible with the expected data. If there is no use case for returning a dict as an iterator, it could even raise a warning if the `parse_response` returns a dict.

## Logs
```
{"type": "LOG", "log": {"level": "INFO", "message": "Starting syncing SourceNativex"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Syncing stream: campaigns "}}
{"type": "LOG", "log": {"level": "ERROR", "message": "Encountered an exception while reading stream campaigns\nTraceback (most recent call last):\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 113, in read\n    yield from self._read_stream(\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 182, in _read_stream\n    for record in record_iterator:\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 282, in _read_full_refresh\n    yield self._as_airbyte_record(configured_stream.stream.name, record)\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 317, in _as_airbyte_record\n    message = AirbyteRecordMessage(stream=stream_name, data=data, emitted_at=now_millis)\n  File \"pydantic/main.py\", line 341, in pydantic.main.BaseModel.__init__\npydantic.error_wrappers.ValidationError: 1 validation error for AirbyteRecordMessage\ndata\n  value is not a valid dict (type=type_error.dict)"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Finished syncing campaigns"}}
{"type": "LOG", "log": {"level": "INFO", "message": "SourceNativex runtimes:\nSyncing stream campaigns 0:00:01.377093"}}
{"type": "LOG", "log": {"level": "FATAL", "message": "1 validation error for AirbyteRecordMessage\ndata\n  value is not a valid dict (type=type_error.dict)\nTraceback (most recent call last):\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/main.py\", line 13, in <module>\n    launch(source, sys.argv[1:])\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 123, in launch\n    for message in source_entrypoint.run(parsed_args):\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 114, in run\n    for message in generator:\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 127, in read\n    raise e\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 113, in read\n    yield from self._read_stream(\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 182, in _read_stream\n    for record in record_iterator:\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 282, in _read_full_refresh\n    yield self._as_airbyte_record(configured_stream.stream.name, record)\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 317, in _as_airbyte_record\n    message = AirbyteRecordMessage(stream=stream_name, data=data, emitted_at=now_millis)\n  File \"pydantic/main.py\", line 341, in pydantic.main.BaseModel.__init__\npydantic.error_wrappers.ValidationError: 1 validation error for AirbyteRecordMessage\ndata\n  value is not a valid dict (type=type_error.dict)"}}
{"type": "TRACE", "trace": {"type": "ERROR", "emitted_at": 1663859942510.5889, "error": {"message": "Something went wrong in the connector. See the logs for more details.", "internal_message": "1 validation error for AirbyteRecordMessage\ndata\n  value is not a valid dict (type=type_error.dict)", "stack_trace": "Traceback (most recent call last):\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/main.py\", line 13, in <module>\n    launch(source, sys.argv[1:])\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 123, in launch\n    for message in source_entrypoint.run(parsed_args):\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py\", line 114, in run\n    for message in generator:\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 127, in read\n    raise e\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 113, in read\n    yield from self._read_stream(\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 182, in _read_stream\n    for record in record_iterator:\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 282, in _read_full_refresh\n    yield self._as_airbyte_record(configured_stream.stream.name, record)\n  File \"/Users/.../airbyte_trial/airbyte-integrations/connectors/source-nativex/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py\", line 317, in _as_airbyte_record\n    message = AirbyteRecordMessage(stream=stream_name, data=data, emitted_at=now_millis)\n  File \"pydantic/main.py\", line 341, in pydantic.main.BaseModel.__init__\npydantic.error_wrappers.ValidationError: 1 validation error for AirbyteRecordMessage\ndata\n  value is not a valid dict (type=type_error.dict)\n", "failure_type": "system_error"}}}
```

## Steps to Reproduce
1. Return a dict from a connectors `parse_response` method
2. Call `python3 main.py read --config secrets/config.json --catalog catalog.json` on it

## Are you willing to submit a PR?
Sure. From what I can see the method is called from https://github.com/airbytehq/airbyte/blob/bb39b36016ca32baba3a762ee2a56fe3da358bcc/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py#L428
Would it make sense to do a type assert here?

### PR DESCRIPTION
**Pull request by Gitznik**
## What
The change makes it easier to debug when the user returns the response dict directly from the parse_response method in the python cdk, instead of yielding it. Before this change, it leads to a Pydantic validation error that does not help pointing you in the right direction: 
`pydantic.main.BaseModel.__init__\npydantic.error_wrappers.ValidationError: 1 validation error for AirbyteRecordMessage\ndata\n  value is not a valid dict (type=type_error.dict)`

Closes #17047

## How
Added a validator that runs before the default pydantic validators that checks if the data is a valid dict and if not, raises a ValueError with a pointer to the return type of the parse_response method.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Airbyte Python CDK Improvement</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Unit & integration tests added and passing: 914 passed, 669 warnings in 10.04s. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

Test output is too long to add it here.

============================================================================================================================================== 914 passed, 669 warnings in 10.54s ===============================================================================================================================================

</details>

### CODE REVIEW

**Review by girarda — CHANGES_REQUESTED** [review]
Thank you for your contribution! Two small comments then we should be good to merge

File: airbyte-cdk/python/airbyte_cdk/models/airbyte_protocol.py [hunk 1]
```diff
@@ -35,6 +35,15 @@ class Config:
         description="when the data was emitted from the source. epoch in millisecond.",
     )
 
+    @validator("data", pre=True)
+    def data_is_dict(cls: AirbyteRecordMessage, value: Dict[str, Any]):
+        if isinstance(value, dict):
+            return value
+        raise ValueError("Data object is not a dictionary. "
```
[the remark this answers is not in the document]
Comment (Gitznik) [reply airbyte-cdk/python/airbyte_cdk/models/airbyte_protocol.py:42]: Great point. Added.

**Review by marcosmarxm — APPROVED** [review]

**Comment by marcosmarxm** [comment]
@girarda can you review this again?

**Review by girarda — APPROVED** [review]

**Comment by girarda** [comment]
Love it! Thank you for your contribution @Gitznik ❤️ 

I'll publish the CDK in a follow-up PR

**Pull request merged**